lib/dao.js

```diff
@@ -2687,6 +2687,14 @@ DataAccessObject.replaceById = function(id, data, options, cb) {
 
   Model.notifyObserversOf('before save', context, function(err, ctx) {
     if (err) return cb(err);
+    if (Model.settings.forceId && ctx.instance[pkName] !== id) {
+      var err = new Error ('id property cannot be changed from ' +
+        id + ' to ' + inst[pkName] + ' in \'before save\' operation ' +
+        'hook when forceId is set to true');
+      err.statusCode = 403;
+      process.nextTick(function() { cb(err); });
+      return cb.promise;
```

Comment:
BTW if this code was not in a next tick, then L2689 above would need to defer callback to next tick too.